## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [151]:
import pandas as pd
pd.set_option("display.precision", 2)

df = pd.read_csv('Resources/client_dataset.csv')

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [152]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [153]:
# Use the describe function to gather some basic statistics

df.describe()

client_id  order_id  order_week  order_year  unit_price  unit_cost  \
count   54639.00  5.46e+04    54639.00    54639.00    54639.00   54639.00   
mean    54837.87  5.47e+06       11.36     2022.99      136.27      99.45   
std     25487.44  2.60e+06        7.02        0.08      183.87     133.16   
min     10033.00  1.00e+06        1.00     2022.00        0.01       0.01   
25%     33593.00  3.20e+06        6.00     2023.00       20.80      14.84   
50%     53305.00  5.50e+06       11.00     2023.00       68.31      49.89   
75%     78498.00  7.73e+06       17.00     2023.00      173.16     125.57   
max     99984.00  1.00e+07       52.00     2023.00     1396.23     846.27   

       unit_weight       qty  line_number  
count     54639.00  5.46e+04     54639.00  
mean          5.00  5.70e+02         2.98  
std           5.33  1.88e+04         2.44  
min           0.00  0.00e+00         0.00  
25%           1.45  3.20e+01         1.00  
50%           3.24  6.80e+01         3.00  
75%           6.89  1.70e+02         5.00  
max          46.43  3.96e+06         9.00

In [154]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.isna().sum()

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [155]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [156]:
df['category'].value_counts()

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [157]:
categorydf = df.groupby('category')

categorydf['subcategory'].value_counts()

category     subcategory       
consumables  bathroom supplies     6424
             paper products        5874
             kitchen supplies      5746
             misc                  2891
             pens                  2603
decor        plants                3466
             wall art              3093
             misc                  1395
electronics  telecom                978
             audio visual           732
             cabling                655
             computers              467
furniture    tables                4540
             stands                4201
             seating               2664
             desks                  510
software     accounting            3756
             project management    3522
             logistics             1122
Name: count, dtype: int64

In [158]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [159]:
# For the category with the most entries,
# which subcategory had the most entries?
top_category = df['category'].value_counts().head(1).index[0]

df.loc[df['category'] == top_category, 'subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [160]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [161]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().head(5).index.tolist()

top_clients

[33615, 66037, 46820, 38378, 24741]

In [162]:
# How many total units (the qty column) did the
# client with the most entries order order?

df.loc[df['client_id'] == top_clients[0], 'qty'].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [163]:
df['subtotal'] = df['unit_price'] * df['qty']

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [164]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['subtotal'] = df['unit_price'] * df['qty']

df[['unit_price', 'qty', 'subtotal']].head(2)

unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95

In [165]:

df['total_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)

df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head()

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [166]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

df['total_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)

df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [167]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['line_price'] = round((df['subtotal'] + df['shipping_price']) * 1.0925, 2)

df[['subtotal', 'shipping_price', 'line_price']].head(3)

subtotal  shipping_price  line_price
0  115164.00         5512.50   131839.08
1     523.95          312.90      914.26
2     527.28          458.64     1077.12

In [168]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_cost'] = round((df['unit_cost'] * df['qty']) + df['shipping_price'], 2)

df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number   subtotal  total_weight  shipping_price  line_price  line_cost  
0            1  115164.00        787.50         5512.50   131839.08   85597.05  
1            0     523.95         31.29          312.90      914.26     629.79  
2            6     527.28         65.52          458.64     1077.12     765.18  

[3 rows x 23 columns]

In [169]:
# Create a column for the profit of
# each line using line cost and line price

df['line_profit'] = round(df['line_price'] - df['line_cost'], 2)

df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

    subtotal  total_weight  shipping_price  line_price  line_cost  line_profit  
0  115164.00        787.50         5512.50   131839.08   85597.05     46242.03  
1     523.95         31.29          312.90      914.26     629.79       284.47  
2     527.28         65.52          458.64     1077.12     765.18       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [170]:
# df.loc[df['email'] == 'blair.kimberly.6618@nelson.biz'].sort_values('order_id').head(20)

df.loc[df['order_id'] == 2742071].head()
# df.loc[df['order_id'] == 2173913].head()
# df.loc[df['order_id'] == 6128929].head()

# df.loc[df['email'] == '

first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year  ... unit_cost  \
3446    2742071  2023-02-27           9        2023  ...      3.05   
6287    2742071  2023-02-27           9        2023  ...    407.68   
20623   2742071  2023-02-27           9        2023  ...     80.78   
23840   2742071  2023-02-27           9        2023  ...     35.41   
24671   2742071  2023-02-27           9        2023  ...     54.86   

      unit_weight  qty  line_number   subtotal  total_weight  shipping_price  \
3446        10.46   77            5     430.43        805.42         5637.94   
6287         2.60  260            0  119652.00        676.00         4732.00   
20623        2.27   12            3    1621.44         27.24          272.40   
23840        0.54   22            1    1046.98         11.88          118.80   
24671        1.57   24            2    1432.56         37.68          376.80   

       line_price  line_cost  line_profit  
3446      6629.69    5872.79       756.90  
6287    135889.52  110728.80     25160.72  
20623     2069.02    1241.76       827.26  
23840     1273.61     897.82       375.79  
24671     1976.73    1693.44       283.29  

[5 rows x 24 columns]

In [171]:
# Check your work using the totals above
# The email is not the same for the orders but this follows what was persctibed in the instructions

orders = [2742071, 2173913, 6128929]

def get_order_total(order_id):
    return df.loc[df['order_id'] == order_id, 'line_price'].sum()

print('You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.')
for order in orders:
    print(f"Order ID {order} had a total price of ${get_order_total(order):,.2f}.")

You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.
Order ID 2742071 had a total price of $152,811.89.
Order ID 2173913 had a total price of $162,388.71.
Order ID 6128929 had a total price of $923,441.25.


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [172]:
for client in top_clients:
    print(f"Client {client} had a total order price of ${df.loc[df['client_id'] == client, 'line_price'].sum():,.2f}.")

Client 33615 had a total order price of $8,377,308.52.
Client 66037 had a total order price of $10,259,514.79.
Client 46820 had a total order price of $9,743,794.36.
Client 38378 had a total order price of $12,906,550.87.
Client 24741 had a total order price of $82,268,892.02.


In [173]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
for client in top_clients:
    print(f"{client}: ${df.loc[df['client_id'] == client, 'line_price'].sum():,.2f}.")


33615: $8,377,308.52.
66037: $10,259,514.79.
46820: $9,743,794.36.
38378: $12,906,550.87.
24741: $82,268,892.02.


In [175]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary = pd.DataFrame(columns=['client_id', 'qty', 'total_shipping', 'total_price', 'total_cost', 'total_profit'])


for client in top_clients:
    total_units = int(df.loc[df['client_id'] == client, 'qty'].sum())
    total_shipping = round(df.loc[df['client_id'] == client, 'shipping_price'].sum(), 2)
    total_price = round(df.loc[df['client_id'] == client, 'line_price'].sum(), 2)
    total_cost = round(df.loc[df['client_id'] == client, 'line_cost'].sum(), 2)
    total_profit = round(df.loc[df['client_id'] == client, 'line_profit'].sum(), 2)
    
    summary.loc[len(summary)] = [int(client), total_units, total_shipping, total_price, total_cost, total_profit]

summary = summary.sort_values('total_profit', ascending=False)
summary = summary.astype({'client_id': 'int64', 'qty': 'int64', 'total_shipping': 'string', 'total_price': 'string', 'total_cost': 'string', 'total_profit': 'string'})

summary.head()


client_id     qty total_shipping  total_price   total_cost total_profit
4      24741  239862     5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.4  12906550.87   9634720.98   3271829.89
1      66037   43018     1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768     1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313     1828984.89   8377308.52   6175313.91   2201994.61

In [176]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

summary = summary.astype({'client_id': 'int64', 'qty': 'int64', 'total_shipping': 'string', 'total_price': 'string', 'total_cost': 'string', 'total_profit': 'string'})

summary['total_shipping'] = summary['total_shipping'].apply(lambda x: f"${(float(x) / 1000000):0.2f}M")
summary['total_price'] = summary['total_price'].apply(lambda x: f"${(float(x) / 1000000):0.2f}M")
summary['total_cost'] = summary['total_cost'].apply(lambda x: f"${(float(x) / 1000000):0.2f}M")
summary['total_profit'] = summary['total_profit'].apply(lambda x: f"${(float(x) / 1000000):0.2f}M")


summary = summary.rename(columns={'client_id': 'Client ID', 'qty': 'Total Units Purchased', 'total_shipping': 'Total Shipping Cost (Millions)', 'total_price': 'Total Price (Millions)', 'total_cost': 'Total Cost (Millions)', 'total_profit': 'Total Profit (Millions)'})

summary.head()

Client ID  Total Units Purchased Total Shipping Cost (Millions)  \
4      24741                 239862                         $5.13M   
3      38378                  73667                         $3.43M   
1      66037                  43018                         $1.40M   
2      46820                  75768                         $1.60M   
0      33615                  64313                         $1.83M   

  Total Price (Millions) Total Cost (Millions) Total Profit (Millions)  
4                $82.27M               $45.69M                 $36.58M  
3                $12.91M                $9.63M                  $3.27M  
1                $10.26M                $7.00M                  $3.26M  
2                 $9.74M                $7.01M                  $2.74M  
0                 $8.38M                $6.18M                  $2.20M

In [177]:
# Sort the updated data by "Total Profit" form highest to lowest

summary = summary.sort_values('Total Profit (Millions)', ascending=False)

summary.head()


Client ID  Total Units Purchased Total Shipping Cost (Millions)  \
4      24741                 239862                         $5.13M   
3      38378                  73667                         $3.43M   
1      66037                  43018                         $1.40M   
2      46820                  75768                         $1.60M   
0      33615                  64313                         $1.83M   

  Total Price (Millions) Total Cost (Millions) Total Profit (Millions)  
4                $82.27M               $45.69M                 $36.58M  
3                $12.91M                $9.63M                  $3.27M  
1                $10.26M                $7.00M                  $3.26M  
2                 $9.74M                $7.01M                  $2.74M  
0                 $8.38M                $6.18M                  $2.20M

In [178]:
df.dtypes

first              object
last               object
job                object
phone              object
email              object
client_id           int64
order_id            int64
order_date         object
order_week          int64
order_year          int64
item_id            object
category           object
subcategory        object
unit_price        float64
unit_cost         float64
unit_weight       float64
qty                 int64
line_number         int64
subtotal          float64
total_weight      float64
shipping_price    float64
line_price        float64
line_cost         float64
line_profit       float64
dtype: object